contrib/mesos/pkg/election/etcd_master_test.go

```diff
@@ -58,41 +56,71 @@ func TestEtcdMasterNoOther(t *testing.T) {
 	w.Stop()
 }
 
-func TestEtcdMasterNoOtherThenConflict(t *testing.T) {
-	path := "foo"
-	e := tools.NewFakeEtcdClient(t)
-	e.TestIndex = true
-	// Ok, so we set up a chain of responses from etcd:
-	//   1) Nothing there
-	//   2) conflict (someone else wrote)
-	//   3) new value (the data they wrote)
-	empty := tools.EtcdResponseWithError{
-		R: &etcd.Response{
-			Node: nil,
-		},
-		E: &etcd.EtcdError{
-			ErrorCode: tools.EtcdErrorCodeNotFound,
-		},
-	}
-	empty.N = &tools.EtcdResponseWithError{
-		R: &etcd.Response{},
-		E: &etcd.EtcdError{
-			ErrorCode: tools.EtcdErrorCodeNodeExist,
-		},
+// MockClient is implementing the test scenario for TestEtcdMasterNoOtherThenConflict:
+// 1. After first Get() is setting "baz" for the key "foo".
+// 2. Then is expecting Create() call.
+// 3. Then is expecting another Get() call.
+type MockClient struct {
+	client tools.EtcdClient
+	t      *testing.T
+	step   int
+}
+
+func (m *MockClient) GetCluster() []string {
+	return m.client.GetCluster()
+}
+
+func (m *MockClient) Get(key string, sort, recursive bool) (*etcd.Response, error) {
+	response, err := m.client.Get(key, sort, recursive)
+	if m.step == 0 {
+		if _, err := m.client.Set("foo", "baz", 0); err != nil {
+			m.t.Errorf("unexpected error: %v", err)
+		}
+	} else if m.step != 2 {
+		m.t.Errorf("unexpected Get() call: %d", m.step)
 	}
-	empty.N.N = &tools.EtcdResponseWithError{
-		R: &etcd.Response{
-			Node: &etcd.Node{
-				Value: "baz",
-			},
-		},
+	m.step++
+	return response, err
+}
+
+func (m *MockClient) Set(key, value string, ttl uint64) (*etcd.Response, error) {
+	return m.client.Set(key, value, ttl)
+}
+
+func (m *MockClient) Create(key, value string, ttl uint64) (*etcd.Response, error) {
+	if m.step != 1 {
+		m.t.Errorf("unexpected Create() call: %d", m.step)
 	}
-	e.Data["foo"] = empty
-	master := NewEtcdMasterElector(e)
+	m.step++
+	return m.client.Create(key, value, ttl)
+}
+
+func (m *MockClient) CompareAndSwap(key, value string, ttl uint64, prevValue string, prevIndex uint64) (*etcd.Response, error) {
+	return m.client.CompareAndSwap(key, value, ttl, prevValue, prevIndex)
+}
+
+func (m *MockClient) Delete(key string, recursive bool) (*etcd.Response, error) {
+	return m.client.Delete(key, recursive)
+}
+
+func (m *MockClient) Watch(prefix string, waitIndex uint64, recursive bool, receiver chan *etcd.Response, stop chan bool) (*etcd.Response, error) {
+	return m.client.Watch(prefix, waitIndex, recursive, receiver, stop)
+}
+
+func TestEtcdMasterNoOtherThenConflict(t *testing.T) {
+	server := etcdtesting.NewEtcdTestClientServer(t)
+	defer server.Terminate(t)
+
+	client := &MockClient{client: server.Client, t: t}
+	path := "foo"
+	master := NewEtcdMasterElector(client)
 	w := master.Elect(path, "bar")
 	result := <-w.ResultChan()
-	if result.Type != watch.Modified || result.Object.(Master) != "bar" {
+	if result.Type != watch.Modified || result.Object.(Master) != "baz" {
```

Comment:
This test previously didn't work - there is a bug in FakeEtcdClient so it didn't correcly processed response.N fields.